GDrive Implementation

nbey opened the pull request.

nbey committed:
Remove unneeded test data

nbey committed:
Add google apps domain to site environment variable

nbey committed:
Include Google Apps API script

nbey committed:
Enable attachment button

nbey committed:
Concisely rename method for adding link attachments

nbey committed:
Fire event for google attachment button click

nbey committed:
Load google api script with current protocol

nbey committed:
Allow teacher to authenticate with google account

nbey committed:
Fix 'fireEvent' call for google attachments

nbey committed:
Disable google attachments for existing tasks

nbey committed:
Check if token exists before authenticating

nbey committed:
Verify authenticated user is within configured domain

nbey committed:
Update attachment model

nbey committed:
Configure google attachment settings menu to store selections

nbey committed:
Update `GoogleDocument` class

nbey committed:
Add 'Share' and 'File' fields to attachment model

nbey committed:
Re-configure google attachment sharing preferences

nbey committed:
Update text for 'save' button to 'publish'

nbey committed:
Show confirmation dialog when attaching google drive files

nbey committed:
Add google file details to respective field

nbey committed:
Add required Google classes

nbey committed:
Add share column to model

nbey committed:
Add file relationship configuration

themightychris commented on php-classes/Slate/CBL/Tasks/Attachments/GoogleDocument.php:
```diff
@@ -2,16 +2,33 @@
-        'ExternalID'
+        'FileID' => 'uint',
+        'RevisionID',
+        'Share' => [
```
@nbey rename this to `ShareMethod`

themightychris commented on php-classes/Slate/CBL/Tasks/Attachments/GoogleDocument.php:
```diff
@@ -2,16 +2,33 @@
 namespace Slate\CBL\Tasks\Attachments;
 
-class GoogleDocument extends AbstractTaskAttachment
+class GoogleDocument extends Link
```
@nbey why should this extend link? A google drive file doesn't have a URL field to store distinctly from its ID. It seems to me that it should continue extending `AbstractTaskAttachment`

themightychris replied:
@nbey also, this class should be renamed to `GoogleDriveFile`

themightychris commented on php-classes/Slate/CBL/Tasks/Attachments/GoogleDocument.php:
```diff
@@ -2,16 +2,33 @@
+    public static $relationships = [
+        'File' => [
+            'type' => 'one-one',
+            'class' => \Google\File::class
```
@nbey let's call this class `\Google\DriveFile`

themightychris commented on php-classes/Google/API.php:
```diff
@@ -0,0 +1,157 @@
+        ]);
+
+        if (empty($response['access_token'])) {
+            \Debug::dumpVar($response, false, 'api response');
```
@nbey remove this

themightychris commented on php-classes/Google/File.php:
```diff
@@ -0,0 +1,84 @@
+            'values' => ['normal', 'trashed', 'deleted'],
+            'default' => 'normal'
+        ],
+        'ParentDocumentIdentifier' => [
```
@nbey rename to `ParentDriveID`

themightychris commented on php-classes/Google/File.php:
```diff
@@ -0,0 +1,84 @@
+    public $details;
+
+    public static $fields = [
+        'ExternalIdentifier',
```
@nbey rename to `DriveID`

themightychris replied:
@nbey add `'unique' => true` option

themightychris commented on sencha-workspace/packages/slate-cbl/src/model/tasks/Attachment.js:
```diff
@@ -13,46 +13,60 @@ Ext.define('Slate.cbl.model.tasks.Attachment', {
         {
-            name: "URL",
-            type: "string",
+            name: 'RevisionID',
```
@nbey exclude the RevisionID field from client-side models

themightychris commented on sencha-workspace/packages/slate-cbl/src/view/AttachmentsList.js:
```diff
@@ -82,31 +82,46 @@ Ext.define('Slate.cbl.view.AttachmentsList', {
+                            items: [{
+                                boxLabel: 'View Only',
+                                inputValue: 'view-only',
+                                checked: Boolean(record.get('Share') == 'view-only')
```
@nbey did some linter make you use this explicit cast? `==` should always return boolean inherently

themightychris commented on sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:
```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
+                    return me.doOpenFilePicker(token);
+                }
+
+                me.doGoogleAuthentication();
```
@nbey doOpenFilePicker should get passed as a callback to this method rather than being hard-coded into the chain of functions that follows authentication

We're going to want to break out all these functions to a sharable singleton class in the slate-cbl package so that other UIs (like the student submission UI we'll implement by the end of this project) can use the same code. Think about how to implement a clean, resuable set of methods there that could be called from here and chained together here through callback parameters

themightychris commented on sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:
```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
+                fields: 'user'
+            }
+        }).
+        then(_verifyUserIsWithinDomain);
```
@nbey inline the function instead of storing it in a variable to keep things easier to follow unless there are multiple paths to that function that necessitate making it reusable

themightychris commented on sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:
```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
+            }).
+            then(_getGoogleFileOwner).
+            then(_getGoogleFileOwnerEmail).
+            then(_doVerifyGoogleFileOwner);
```
@nbey again, inline functions if they're not being reused so we can see all the code in sequence

themightychris commented on sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:
```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
+            console.info('loaded user', response.result);
+
+            if (!response.result || !response.result.user || !response.result.user.emailAddress.match(googleAppsEmailRegex)) {
+                return _userIsNotWithinDomain('Account must be under the correct domain. ' + googleAppsDomain);
```
@nbey reformat this message so it reads `'Please sign in with an account under slatedemo.com'`

themightychris replied:
@nbey after this dialog is shown, clicking "attachment" again jumps right back to it without letting the user sign in again. See if there is a client api call you can make to "deauth" the user from the current js api session, ideally without signing them out of google. If you reload the page you can pick an account again so there should be a call in the js API to achieve the same thing

themightychris replied:
looks like reloading the page doesn't reset it either, we need to find a js api call we can make to reset auth, ideally without logging out the user's session

themightychris commented on sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js:
```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
+            tokenExpiration = new Date(authResult.expires_at * 1000);
+
+            Ext.util.Cookies.set('googleAppsToken', authResult.access_token, tokenExpiration, '/');
+            Ext.util.Cookies.set('googleAppsPermissionId', response.result.user.permissionId, tokenExpiration, '/');
```
@nbey I don't think we should store the permissionId in a cookie, only the token. If the user comes back with a token, we should hit whatever API provides us the permissionId first to both verify that the token is valid and re-obtain the permission id for the current page load

nbey committed:
Remove debug statement

nbey committed:
Rename `GoogleDocument` class to `GoogleDriveFile`

nbey committed:
Revert  class extension to `AbstractTaskAttachment`

nbey committed:
Rename column to `ShareMethod`

nbey committed:
Update class name and references to `DriveFile`

nbey committed:
Rename ID column and make unique

nbey committed:
Rename parent ID column

nbey committed:
Update `GoogleDriveFile` reference

nbey committed:
Update drive file identifier column reference

nbey committed:
Validate DriveFile model required fields

nbey committed:
Update Drive file identifier column name

nbey committed:
Update GoogleDriveFile class reference

nbey committed:
Remove unneeded attachment model fields

nbey committed:
Fix google file share method preferences menu

nbey committed:
Use Google utility class to facilitate common google api methods

nbey committed:
Add method for getting required drive file permissions

nbey committed:
Move code for creating task attachments to `applyRecordDelta`

nbey committed:
Attempt to use current session user email if `OwnerEmail` isn't set

nbey committed:
Check user file permissions when collaborate share method is selected

nbey committed:
Use shortcut method for managing tokens

nbey committed:
Remove unneeded code and fix typos

nbey committed:
Reset access token when authenticated user is cleared

nbey committed:
Fire `sharemethodchange` event when sharing preference changes

nbey committed:
Fix jslint code format issues

nbey committed:
Add required class use statements

nbey committed:
Fix google drive file clone workflow

nbey committed:
Verify re-used access tokens

nbey committed:
Update google file permissions after saving tasks

nbey committed:
Commit required gDrive SiteEnvironment variables

nbey committed:
Update error messages text

nbey committed:
Remove unneeded code

themightychris commented on php-classes/Google/DriveFile.php:
```diff
@@ -0,0 +1,108 @@
+
+use \Slate\CBL\Tasks\Attachments\GoogleDriveFile;
+use \Slate\CBL\Tasks\Task;
+use \Slate\CBL\Tasks\StudentTask;
```
@nbey code in the `Google` namespace should not be relying on code in the `Slate` namespace. Any `Slate`-dependent code should be in the `Slate` namespace. It doesn't look like any of these are actually used in this file anyway.

Also, `use` statements don't need to start with `\`, they are inherently absolute.

themightychris commented on php-classes/Google/DriveFile.php:
```diff
@@ -0,0 +1,108 @@
+
+        if (!\Validators::string($File->Title)) {
+            $validator->addError('Title', 'Title is missing or invalid.');
+        }
```
@nbey put `Validators`, `Exception`, and `RecordValidator` in `use` statements at the top and drop the leading `\` throughout the class body.

themightychris commented on php-classes/Google/DriveFile.php:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    public static function validateTitle(\RecordValidator $validator, \Google\DriveFile $File)
```
@nbey `DriveFile` can be referenced just as `DriveFile` since it's in the same namespace. Further I think you could just use `self` or `static` here

nbey committed:
Include file attachment data

nbey committed:
Use file attachment title when available

nbey committed:
Fix google file link urls

nbey committed:
Include file attachment data

nbey committed:
Allow for token to be set via options config

nbey committed:
Require user when retrieving access token

nbey committed:
Remove unneeded use statements

nbey committed:
Set details when retrieving google file details

nbey committed:
Update class type hints to `self`

nbey committed:
Use token when creating permissions

nbey committed:
Add method for duplicating google files

nbey committed:
Add `ParentAttachment` column and relationship

nbey committed:
Return permission type if specified

nbey committed:
Fix required permissions list for duplicate files

nbey committed:
Add dynamic fields for file and parent attachment

nbey committed:
Add support for retrieving required permissions for phantom attachments

nbey committed:
Duplicate google file attachments on task record save

nbey committed:
Add classes to use statements

nbey committed:
Include attachments file data

nbey committed:
Prevent non-phantom google drive file changes

nbey committed:
Implement google API for adding attachments to student tasks

themightychris committed:
Update git mapping path for DriveFile class

nbey committed:
Remove skills assignment logic

nbey committed:
Remove unneeded code

nbey committed:
Select existing drive file record before creating one

nbey committed:
Remove unneeded debug log statements

nbey committed:
Implement user access token caching

nbey committed:
Add support for email domain validations

nbey committed:
Return from method if file owner email is not set

nbey committed:
Update API request method parameters
- Leverage access token caching from API class
- Supply exception error code for file detail lookups that result in 404: file not found
- Remove token argument from `duplicate` / `createPermission` methods
- Add `untrash` method
- Add `updateGoogleFileDetails` method for applying google file details data to model
- Remove trashed shared drive files from the trash
- Send email notification for trashed/deleted drive files

nbey committed:
WIP: Implement google drive user changes monitor

nbey committed:
Remove unneeded code for fetching and reusing access tokens

nbey committed:
Use `toString` method for email contact models

nbey committed:
Remove unused variable

nbey committed:
Add `Google\Drive` namespace to git config

themightychris committed:
Delete obsolete comment

themightychris committed:
Improve token usage options; unify cache key generation

themightychris committed:
Merge branch 'develop' into features/google-drive

* develop:
  Wrap `getCurrentLevelForStudent` logic in a try block

nbey committed:
Update `RevisionID` column name to `FileRevisionID`

- prevent conflicts with `VersionedRecord` `RevisionID` column

nbey committed:
Add migration script for `GoogleDriveFile` attachment columns

nbey committed:
Store google api credentials and load script

nbey committed:
Allow adding attachments to existing tasks

nbey committed:
Catch any authentication or file picker errors

nbey committed:
Remove unneccesary caching of permissionId

nbey committed:
Use defined `me` scope

themightychris commented on php-classes/Slate/CBL/Tasks/TasksRequestHandler.php:
```diff
@@ -377,7 +377,7 @@ public static function updateGoogleFilePermissions(Task $Record)
                                 'Context' => $StudentTask,
                                 'ParentAttachment' => $attachment,
-                                'RevisionID' => 1
+                                'FileRevisionID' => 1
```
@nbey do we not have access to the actual revision ID? Does `1` pass string validation?

nbey committed:
Use Google\API class variables for storing credentials

nbey committed:
Use the file owner email when creating permissions

nbey committed:
Update attachment creation logic

nbey committed:
Fix teacher permissions on student google drive attachments

nbey committed:
Move logic for updating google drive file permissions

nbey committed:
Replace deprecated gapi `authorize` method

nbey committed:
Fix Google drive file duplication workflow

- Update `\Google\DriveFile::duplicate` method to return a duplicated `\Google\Drive` file object.

nbey committed:
Update language in clone confirmation window

nbey committed:
Mask task editor window during publish

nbey committed:
Return file owner email when cloning files

nbey committed:
Disable google attachment button if not configured

nbey committed:
Handle google sign-in error codes

nbey committed:
Use string value for `FileRevisionID` column

themightychris committed:
Trim whitespace

themightychris committed:
Tweak whitespace and capitalization

nbey committed:
Handle 'channelIdNotUnique` errors when monitoring users

- `channelIdNotUnique` error means that the user monitor is already configured and so it should be ignored.

nbey committed:
Move `nightly-maintenance` event handlers

- Move `nightly-maintenance` event handlers from `Slate\CBL` namespace to `Emergence\Site`
- Prefix cbl related event handlers with `cbl`

nbey committed:
Add method for monitoring all teachers

nbey committed:
Add google drive change monitor webhook address

nbey committed:
Update git tree mappings

- Move and add `nightly-maintenance` event handlers from `Slate\CBL` namespace to `Emergence\Site` namespace
- Add Google Drive change monitor webhook path in `site-root`

nbey committed:
Add `Status` field to attachments model

nbey committed:
Add `Status` field to AbstractTaskAttachment class

nbey committed:
Implement editing of Attachment status in the backend

nbey committed:
Add migration script for AbstractTaskAttachment `Status` column

nbey committed:
Implement task attachment status moderation

nbey committed:
Allow administrators to browse google drive files

nbey committed:
Allow updating of attachment status

nbey committed:
Implement config variables for editing attachments

- Enable (soft) deletion of attachments
- Configure share method preferences

nbey committed:
Update task details attachments UI in `SlateTasksStudent`

- Include `duplicate` and `collaborate` Google drive attachments in student attachments section (hide them from teacher attachment section)

-  Disable share method preferences

- Only allow (soft) deletion of attachments created by the student

nbey committed:
Fix typo

nbey committed:
Prevent API error when term is not configured

nbey committed:
Update method name

nbey committed:
Update student tasks during `applyRecordDelta`

Move logic for assigning student tasks to `applyRecordDelta` method

nbey committed:
Reimplement google drive file permissions workflow

- Use an event handler to sync google drive file permissions on new tasks
- Use `onRecordSaved` method to update google drive file permissions for new student assignments

nbey committed:
Mask all components during requests

nbey committed:
Remove dump statement

nbey committed:
Sync permissions for student attached drive files

nbey committed:
Declare variable before usage

nbey committed:
Mask task details panel during submission

themightychris commented on php-classes/Google/Drive/UserChangeMonitorRequestHandler.php:
```diff
@@ -0,0 +1,163 @@
+        $requestHeaders = getallheaders();
+
+        $channelId = $requestHeaders['X-Goog-Channel-Id'];
+        $resourceState = $requestHeaders['X-Goog-Resource-State'];
```
@nbey why all this reformatting of all the headers instead of just
```
$channelId = $_SERVER['HTTP_X_GOOGL_CHANNEL_ID'];
$resourceState = $_SERVER['HTTP_X_GOOG_RESOURCE_STATE'];
```

themightychris commented on php-classes/Slate/CBL/Tasks/StudentTasksRequestHandler.php:
```diff
@@ -162,75 +165,46 @@ public static function handleStudentTaskSubmissionRequest()
-            $attachmentIds = [];
-            $attachments = [];
-            $defaultAttachmentClass = AbstractTaskAttachment::$defaultClass;
+        $defaultAttachmentClass = AbstractTaskAttachment::class;
```
@nbey `AbstractTaskAttachment` isn't actually declared as an `abstract` class (although it should be or should be named something else). If it really is an abstract class, we can't have an instance of it, and so it shouldn't be a default. An attachment without class=Link or class=GoogleDriveFile is invalid and should generate a validation error or exception, and the abstract class should be actually declared as such so that any attempt to instantiate one generates a proper error instead of just silent weirdness

themightychris commented on php-classes/Slate/CBL/Tasks/StudentTasksRequestHandler.php:
```diff
@@ -162,75 +165,46 @@ public static function handleStudentTaskSubmissionRequest()
-                    ], true);
+                if ($Attachment instanceof Attachments\GoogleDriveFile) {
+                    if (!$Attachment->File) {
+                        if (!$File = \Google\DriveFile::getByWhere(['DriveID' => $attachmentData['File']['DriveID']])) {
```
@nbey `getByField` would be more appropriate here and get to benefit from caching

themightychris committed:
Let API wrapper prepend host

themightychris committed:
Trim trailing whitespace

nbey committed:
Fix google drive image links

nbey committed:
Manage google drive errors better

nbey committed:
Remove unused variable

nbey committed:
Send `SectionID` with all save requests

nbey committed:
Add section relationship validator

nbey committed:
Declare attachments variable

nbey committed:
Remove google drive attachment permission syncing logic

nbey committed:
Use $_GET param to set `SectionID`

nbey committed:
Store permission type in memory

- Only allow upgrade of permissions

nbey committed:
Update permissions logic

- Give teachers writer permission on all student attachments
- Better prevent duplicate cloned attachments
- Remove `syncPermissions` method

nbey committed:
Reconfigure google drive permission syncing logic

- Trigger google drive file permission pushes via event-handlers

nbey committed:
Move attachment event handler path

nbey committed:
Use `getByField` method

nbey committed:
Remove unneeded `getallheaders` method

nbey committed:
Declare `AbstractTaskAttachment` class as abstract

- Use `AbstractTaskAttachment::$defaultClass` as the default class

themightychris commented on php-classes/Slate/CBL/Tasks/TasksRequestHandler.php:
```diff
@@ -163,11 +164,74 @@ public static function handleTaskAssigneesRequest(ActiveRecord $Record)
+                if (!$StudentTask = StudentTask::getByWhere(['StudentID' => $assigneeId, 'TaskID' => $Record->ID])) {
+                    $StudentTask = StudentTask::create([
+                        'StudentID' => $assigneeId,
+                        'SectionID' => $_GET['SectionID'],
```
@nbey Why does this need to come in through the side as a GET param? If it's part of the record it should come in from `$requestData` with the record's other fields. If it's part of the saved record, it seems like the client code could just as easily declare it as a model field and set it before calling `record.save()` as easily as attaching custom request-level params to the save operation?

nbey committed:
Fix typo

nbey committed:
Fix method name typo

nbey committed:
Use `getByFiedl` method

nbey committed:
Fix typo

ryon committed:
Fix modal form control alignments

ryon committed:
Add default case to satisfy linter

ryon committed:
Use showBy() instead of showAt() for ShareMethod menu

ryon committed:
Use CheckItems w/ icons for ShareMethod menu

ryon committed:
Let AttachmentsList items show current ShareMethod instead of gear icon

nbey committed:
Add Emergence Mailer to use statement

nbey committed:
Remove leading backslash from classes

nbey committed:
Revert back to supplying `CourseSectionID` along with Task data

nbey committed:
Remove class namespace

nbey committed:
Convert functions stored in variables to inline

- Improve error handling
- Remove unneccesary escape character

nbey committed:
Add SectionID field to Task model

nbey committed:
Update file cloning workflow

Instead of giving a user read access, and then cloning a file as them, clone the file as the owner, and transfer ownership to the user.

nbey committed:
Reimplement share method menu

- Re-use menus for better performance
- Replace  'click' event with 'tap' event on attachment list items

nbey committed:
Add config for administrative google drive notifications

nbey committed:
Handle errors caused by revisionless drive files

Google drive files without any revisions will throw an error -- ignore the error and add the attachment without the `FileRevisionID` being set.

nbey committed:
Set `FileRevisionID` field default to null and update validator

nbey committed:
Fix duplicate attachment permission issue

Update code so that permissions are only granted on drive files where `ShareMethod` is set to either `collaborate` or `view-only`settings
